passes/techmap/clock_gate.cc

```diff
@@ -0,0 +1,205 @@
+#include "kernel/yosys.h"
+#include "kernel/ff.h"
+#include <optional>
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+struct ClockGateCell {
+	std::string name;
+	std::string ce_pin;
+	std::string clk_in_pin;
+	std::string clk_out_pin;
+
+	ClockGateCell(std::string& name, std::string& str) : name(RTLIL::escape_id(name)) {
+		char delimiter = ':';
+		size_t pos1 = str.find(delimiter);
+		if (pos1 == std::string::npos)
+			log_assert(false && "Not enough ports in descriptor string");
+		size_t pos2 = str.find(delimiter, pos1 + 1);
+		if (pos2 == std::string::npos)
+			log_assert(false && "Not enough ports in descriptor string");
+		size_t pos3 = str.find(delimiter, pos2 + 1);
+		if (pos3 != std::string::npos)
+			log_assert(false && "Too many ports in descriptor string");
+
+		ce_pin = str.substr(0, pos1);
+		ce_pin = RTLIL::escape_id(ce_pin);
+
+		clk_in_pin = str.substr(pos1 + 1, pos2 - (pos1 + 1));
+		clk_in_pin = RTLIL::escape_id(clk_in_pin);
+
+		clk_out_pin = str.substr(pos2 + 1, str.size() - (pos2 + 1));
+		clk_out_pin = RTLIL::escape_id(clk_out_pin);
+	}
+};
+
+struct ClockgatePass : public Pass {
+	ClockgatePass() : Pass("clock_gate", "extract clock gating out of flip flops") { }
```

Comment:
Let's rename this to `clockgate`: if we look at the list of Yosys commands, underscores are used to group commands together but not really as a word separator